File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.cpp
class MedianFinder {
public:
    vector<int> v;
    int size;
    MedianFinder() {
        size = 0;
    }
    
    void addNum(int num) {
        if(size == 0){
            size++;
            v.push_back(num);
            return;
        }
        for(int i=0;i<size;i++){
            if(v[i] > num){
                size++;
                v.insert(v.begin()+i, num);
                return;
            }
        }
        size++;
        v.push_back(num);
    }
    
    double findMedian() {
        int n = size;
        int mid = n / 2;
        bool odd = n % 2 == 1;
        
        return odd ? (double)v[mid] : ((double)(v[mid] + v[mid-1])/2);
    }
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

File: 0295-find-median-from-data-stream/0295-find-median-from-data-stream.dart
import 'package:collection/collection.dart';
class MedianFinder {
  PriorityQueue<int> small = PriorityQueue<int>();
  PriorityQueue<int> large = PriorityQueue<int>((a, b) => b.compareTo(a));
  MedianFinder() {
  }
  
  void addNum(int num) {
    small.add(num);
    large.add(small.removeFirst());
    if(large.length > small.length){
        small.add(large.removeFirst());
    }
  }
  
  double findMedian() {
    return small.length > large.length ? small.first.toDouble() : (small.first + large.first) / 2;
  }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = MedianFinder();
 * obj.addNum(num);
 * double param2 = obj.findMedian();
 */